pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java

```diff
@@ -340,16 +345,24 @@ public void run() {
                     // process the synchronous results
                     handleResult(currentRecord, result);
                 }
+
+                if (deathException != null) {
```

Comment:
We need to judge `deathException != null` in three places
1. When `source.read` just finished.
2. When `function.process` just finished.
3. When `sink.write` just finished.

This can catch the exception in advance and end the thread.

Reply:
If a fatal exception is thrown from source.read or sink.write, the thread can be interrupted, preventing it from reaching this point.
This handles the corner case for the `function.process`. I have added the comment here to minimize the confusion.

```
// Ideally the current java instance thread will be interrupted when the deathException is set.
// But if the CompletableFuture returned by the Pulsar Function is completed exceptionally(the
// function has invoked the fatal method) before being put into the JavaInstance
// .pendingAsyncRequests, the interrupted exception may be thrown when putting this future to
// JavaInstance.pendingAsyncRequests. The interrupted exception would be caught by the JavaInstance
// and be skipped.
// Therefore, we need to handle this case by checking the deathException here and rethrow it.
```

Reply:
Thanks for your explain, I got it. 

+1